### PR TITLE
test: add benchmarks that set bandwidth and timeout constraints

### PR DESCRIPTION
**Pull request by strib**
So we can easily reproduce timeouts while transferring files over various connection speeds.

Issue: KBFS-1236

### CODE REVIEW

File: libkbfs/block_ops.go [hunk 1]
```diff
@@ -137,6 +152,19 @@ func (b *BlockOpsStandard) Ready(ctx context.Context, md *RootMetadata,
 // Put implements the BlockOps interface for BlockOpsStandard.
 func (b *BlockOpsStandard) Put(ctx context.Context, md *RootMetadata,
 	blockPtr BlockPointer, readyBlockData ReadyBlockData) error {
+	if b.bwKBps > 0 {
+		b.lock.Lock()
+		// Simulate a constrained bserver connection
+		delay := len(readyBlockData.buf) * int(time.Second) / (b.bwKBps * 1024)
```
Comment (jzila) [inline libkbfs/block_ops.go]: This seems to add an _additional_ delay, on top of your actual connection's bandwidth. Is that intended?

Comment (strib) [inline libkbfs/block_ops.go]: Yes.  It's intended for use in tests, where there is 0 delay on the "real" connection.

File: test/benchmark_test.go [hunk 2]
```diff
@@ -122,3 +124,73 @@ func benchmarkReadSeqHoleN(b *testing.B, n int64, mask int64) {
 		),
 	)
 }
+
+func benchmarkWriteWithBandwithHelper(b *testing.B, fileBytes int64,
```
Comment (akalin-keybase) [inline test/benchmark_test.go]: Bandwith -> Bandwidth (here and everywhere else)

File: libkbfs/block_ops.go [hunk 3]
```diff
@@ -14,10 +16,23 @@ import (
...
+
+	lock sync.Mutex
 }
 
 var _ BlockOps = (*BlockOpsStandard)(nil)
 
+// NewBlockOpsStandard constructs a new standard implementation of the
+// BlockOps interface. bwKBps indicates a constrained bandwidth to
+// simulate in kilobytes per second (0 means unconstrained).
+func NewBlockOpsStandard(config Config, bwKBps int) *BlockOpsStandard {
```
Comment (akalin-keybase) [inline libkbfs/block_ops.go]: Can you instead implement that bandwidth constraint as a separate type that embeds a `BlockOps`? Something like:

```
type BlockOpsConstrained struct {
  BlockOps
  lock sync.Mutex
  bwKBps int
}

func ... Put(...) {
  b.Lock()
  time.Sleep(...)
  b.Unlock()
  return b.BlockOps.Put(...)
}
```

File: test/engine_common.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+func maybeSetBw(t testing.TB, config libkbfs.Config, bwKBps int) {
+	if bwKBps > 0 {
+		if _, ok := config.BlockOps().(*libkbfs.BlockOpsStandard); ok {
+			config.SetBlockOps(libkbfs.NewBlockOpsStandard(config, bwKBps))
+			// Looks like we're testing big transfers, so let's do
+			// background flushes.
+			config.SetDoBackgroundFlushes(true)
+		} else {
+			t.Logf("Ignore bandwitdh setting of %d for a non-standard block ops",
```
Comment (akalin-keybase) [inline test/engine_common.go]: bandwitdh -> bandwidth

Comment (akalin-keybase) [inline test/engine_common.go]: If you implement the bandwidth constraint as a delegating `BlockOps` type, then you don't need this case.

File: test/benchmark_test.go [hunk 5]
```diff
@@ -122,3 +124,73 @@ func benchmarkReadSeqHoleN(b *testing.B, n int64, mask int64) {
...
+						sync := j+1 == numWritesPerFile
+						err = cb(pwriteBSSync(name, buf,
+							int64(j)*int64(len(buf)), sync))
+						if err != nil {
+							return err
+						}
+					}
+				}
+				if !warmUp {
+					b.StopTimer()
```
Comment (akalin-keybase) [inline test/benchmark_test.go]: do this as a defer in the previous `!warmUp` block?

**Comment by akalin-keybase** [comment]
first pass

File: libkbfs/block_ops_constrained.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+		BlockOps: delegate,
+		bwKBps:   bwKBps,
+	}
+}
+
+// Put implements the BlockOps interface for BlockOpsConstrained.
+func (b *BlockOpsConstrained) Put(ctx context.Context, md *RootMetadata,
+	blockPtr BlockPointer, readyBlockData ReadyBlockData) error {
+	if b.bwKBps > 0 {
+		b.lock.Lock()
```
Comment (akalin-keybase) [inline libkbfs/block_ops_constrained.go]: Nit: Mind decomping out the locked part in a helper function so you can do `defer Unlock()`?

File: libkbfs/bserver_memory.go [hunk 7]
```diff
@@ -191,8 +191,8 @@ func validateBlockServerPut(
 func (b *BlockServerMemory) Put(ctx context.Context, id BlockID, tlfID TlfID,
 	context BlockContext, buf []byte,
 	serverHalf BlockCryptKeyServerHalf) error {
-	b.log.CDebugf(ctx, "BlockServerMemory.Put id=%s tlfID=%s context=%s",
-		id, tlfID, context)
+	b.log.CDebugf(ctx, "BlockServerMemory.Put id=%s tlfID=%s context=%s size=%d",
```
Comment (akalin-keybase) [inline libkbfs/bserver_memory.go]: Can you change `bserver_disk.go` also?

**Comment by akalin-keybase** [comment]
lgtm after nits

**Pull request merged**